### PR TITLE
Adds ServiceAccount Support

### PR DESCRIPTION
**Pull request by danehans**
Previously, federation-v2 was unable to manage federated service accounts. This PR adds support for the `FederatedServiceAccount`, `FederatedServiceAccountPlacement` and `FederatedServiceAccountOverride` types to support federated service accounts.

### CODE REVIEW

**Comment by danehans** [comment]
@irfanurrehman @onyiny-ang I'm not sure why CI is not passing. `go test -v ./test/e2e -args -ginkgo.v -single-call-timeout=5m` passes for me locally:
```
$ go test -v ./test/e2e -args -ginkgo.v -single-call-timeout=5m
=== RUN   TestE2E
Running Suite: Federation e2e suite
===================================
Random Seed: 1535435357
Will run 12 of 12 specs

Aug 28 05:49:17.335: INFO: Starting a federation of 2 clusters...
<SNIP>
Ran 11 of 12 Specs in 125.337 seconds
SUCCESS! -- 11 Passed | 0 Failed | 0 Pending | 1 Skipped
--- PASS: TestE2E (125.34s)
PASS
ok  	github.com/kubernetes-sigs/federation-v2/test/e2e	125.361s
```

**Comment by font** [comment]
@danehans When you run without the `-kubeconfig` argument, you are running managed e2e tests, which are basically the same thing as integration tests except uses the ginkgo framework. We have filed #187 to track the consolidation of integration with managed tests.

The CI is failing in the unmanaged e2e test with 1 cluster. You'll want to stand up your own clusters and run unmanaged e2e by supplying the kubeconfig argument. See the userguide if you want to setup clusters with minikube.

**Comment by danehans** [comment]
I fixed the issue with my original PR. I was using the wrong `comparisonField` value of the `FederatedTypeConfig`. I now get e2e to pass locally. I don't understand why the `FederatedDeployment` is failing in CI. Is there a way to rerun the test?

**Comment by font** [comment]
@danehans If you click on the *Details* link for the failed travis job it'll take you to the build page. From there, you should see a *Restart build* button on the top right.

File: pkg/apis/core/v1alpha1/federatedserviceaccountoverride_types.go [hunk 1]
```diff
@@ -20,21 +20,21 @@ import (
...
-	// Important: Run "kubebuilder generate" to regenerate code after modifying this file
+	Overrides []FederatedServiceAccountClusterOverride `json:"overrides,omitempty"`
+}
+
+// FederatedServiceAccountClusterOverride defines the overrides for a named cluster
+type FederatedServiceAccountClusterOverride struct {
+	// TODO(marun) Need to ensure that a cluster name only appears
+	// once.  Why can't maps be used so this validation is automatic?
+	ClusterName                  string `json:"clusterName,omitempty"`
+	AutomountServiceAccountToken *bool `json:"automountServiceAccountToken,omitempty"`
```
[the remark this answers is not in the document]
Comment (danehans) [reply pkg/apis/core/v1alpha1/federatedserviceaccountoverride_types.go]: @gyliu513 I don't have a use case. Should I not add the override type for the time being and stick with the template and placement types?

**Comment by danehans** [comment]
@font I have no restart button from the build page. Maybe I don't have the necessary permissions?

File: pkg/apis/core/v1alpha1/federatedserviceaccountoverride_types.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (gyliu513) [reply pkg/apis/core/v1alpha1/federatedserviceaccountoverride_types.go]: I think it is OK to have the override policy here, but just not clear what kind of parameters that we want to override for serviceaccount.

Let us get some comments from @font @marun as well for this.

/cc @marun @font

**Comment by gyliu513** [comment]
Thanks @danehans for the update, the overall PR LGTM now.

We can get more comments from @font and @marun

File: pkg/apis/core/v1alpha1/federatedserviceaccountoverride_types.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (marun) [reply pkg/apis/core/v1alpha1/federatedserviceaccountoverride_types.go]: @danehans I suggest avoiding implementing an override type if overrides aren't required.  They can always be added later, but once past beta removal will imply a deprecation process.

File: config/federatedtypes/serviceaccount.yaml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+apiVersion: core.federation.k8s.io/v1alpha1
+kind: FederatedTypeConfig
+metadata:
+  name: serviceaccounts
+spec:
+  target:
+    version: v1
+    kind: ServiceAccount
+  namespaced: true
+  comparisonField: ResourceVersion
```
Comment (gyliu513) [inline config/federatedtypes/serviceaccount.yaml]: Why here using `ResourceVersion` as the comparisonField? I checked the difference `ResourceVersion` and `Generation` at https://stackoverflow.com/questions/47100389/what-is-the-difference-between-a-resourceversion-and-a-generation/47101418 , but still not clear what is the best practise for this value.

FYI @marun

Comment (gyliu513) [reply config/federatedtypes/serviceaccount.yaml]: @danehans I saw that you have updated this value to `Generation` in follow up commits, can you please show some detail for the guidance of when to use `Generation` and when to use `ResourceVersion`? Thanks.

**Comment by danehans** [comment]
@marun @font @gyliu513 I removed the override and CI is passing. Let me know if anything else is needed to merge this PR. Thanks for all your help!

**Comment by gyliu513** [comment]
LGTM, thanks @danehans for the update!

**Comment by danehans** [comment]
Thanks @gyliu513 for the reviews

**Comment by marun** [comment]
@danehans I'd like to merge #216 before this, hopefully by eod today.

**Pull request merged**